# Steam's videogames platform 👾
## Company's description 📇

Steam is a video game digital distribution service and storefront from Valve. It was launched as a software client in September 2003 to provide game updates automatically for Valve's games, and expanded to distributing third-party titles in late 2005. Steam offers various features, like digital rights management (DRM), game server matchmaking with Valve Anti-Cheat measures, social networking, and game streaming services. Steam client's functions include game update automation, cloud storage for game progress, and community features such as direct messaging, in-game overlay functions and a virtual collectable marketplace.

## Project 🚧

You're working for Ubisoft, a French video game publisher. They'd like to release a new revolutionary videogame! They asked you conduct a global analysis of the games available on Steam's marketplace in order to better understand the videogames ecosystem and today's trends.

## Goals 🎯

The ultimate goal of this project is to understand what factors affect the popularity or sales of a video game. But your boss asked you to take advantage of this opportunity to analyze the video game market globally.

To carry out this project, you will have to adopt different levels of analysis. Your boss gave you a list of examples of questions that would be interesting:

### Analysis at the "macro" level

- Which publisher has released the most games on Steam?

- What are the best rated games?

- Are there years with more releases? Were there more or fewer game releases during the Covid, for example?

- How are the prizes distributed? Are there many games with a discount?

- What are the most represented languages?

- Are there many games prohibited for children under 16/18?

### Genres analysis

- What are the most represented genres?

- Are there any genres that have a better positive/negative review ratio?

- Do some publishers have favorite genres?

- What are the most lucrative genres?

### Platform analysis

- Are most games available on Windows/Mac/Linux instead?

- Do certain genres tend to be preferentially available on certain platforms?

You're free to follow these guidelines, or to choose a different angle of analysis, as long as your analysis reveals relevant and useful information. 🤓

## Scope of this project 🖼️

You'll have to use Databricks and PySpark to conduct this EDA. Particularly, you'll have to use Databrick's visualisation tool to create the visualizations.

The dataset is available in our S3 bucket at the following url: s3://full-stack-bigdata-datasets/Big_Data/Project_Steam/steam_game_output.json.

## Helpers 🦮

To help you achieve this project, here are a few tips that should help you:

- To adopt different levels of analysis, it might be useful to create different dataframes.

- As the dataset is semi-structured with a nested schema, Pyspark's methods such as getField() and explode() may help you.

- There are some text and date fields in this dataset: Pyspark offers utilitary functions to manipulate these types of data efficiently 💡

- You can use agregate functions and groupBy to conduct segmented analysis.

## Deliverable 📬

To complete this project, you should deliver:

- One or several notebooks including data manipulation with PySpark and data visualization with Databrick's dashboarding tool.

- To make sure the jury can view all the visualizations, please use the "publish" button on Databricks notebooks to create a public url where a copy of your notebook will be available.

- While using the "publish" button, Databricks may tell you that your notebook's size exceeds the maximal size allowed. If this happens, just split your notebook in several notebooks.

- Please copy-paste the link(s) to your published notebooks into your Github repo such that the jury can access it easily. 😌
  

In [0]:
sc = spark.sparkContext

from pyspark.sql import functions as F # This will load the class where spark sql functions are contained
from pyspark.sql import Row # this will let us manipulate rows with spark sql
from pyspark.sql.types import *

import pandas as pd

In [0]:
filepath = 's3://full-stack-bigdata-datasets/Big_Data/Project_Steam/steam_game_output.json'
steam = spark.read.json(filepath)

In [0]:
steam.printSchema()

root
 |-- data: struct (nullable = true)
 |    |-- appid: long (nullable = true)
 |    |-- categories: array (nullable = true)
 |    |    |-- element: string (containsNull = true)
 |    |-- ccu: long (nullable = true)
 |    |-- developer: string (nullable = true)
 |    |-- discount: string (nullable = true)
 |    |-- genre: string (nullable = true)
 |    |-- header_image: string (nullable = true)
 |    |-- initialprice: string (nullable = true)
 |    |-- languages: string (nullable = true)
 |    |-- name: string (nullable = true)
 |    |-- negative: long (nullable = true)
 |    |-- owners: string (nullable = true)
 |    |-- platforms: struct (nullable = true)
 |    |    |-- linux: boolean (nullable = true)
 |    |    |-- mac: boolean (nullable = true)
 |    |    |-- windows: boolean (nullable = true)
 |    |-- positive: long (nullable = true)
 |    |-- price: string (nullable = true)
 |    |-- publisher: string (nullable = true)
 |    |-- release_date: string (nullable = true)
 |    |-

In [0]:
id_appid = steam.withColumn("are_equal", F.col("data.appid") == F.col("id"))

id_appid.select(F.col("are_equal")).distinct().show()

+---------+
|are_equal|
+---------+
|     true|
+---------+



The "id" column can be dropped since it is identical to the "appid" nested column in "data".

In [0]:
steam = steam.drop("id")
steam.printSchema()

root
 |-- data: struct (nullable = true)
 |    |-- appid: long (nullable = true)
 |    |-- categories: array (nullable = true)
 |    |    |-- element: string (containsNull = true)
 |    |-- ccu: long (nullable = true)
 |    |-- developer: string (nullable = true)
 |    |-- discount: string (nullable = true)
 |    |-- genre: string (nullable = true)
 |    |-- header_image: string (nullable = true)
 |    |-- initialprice: string (nullable = true)
 |    |-- languages: string (nullable = true)
 |    |-- name: string (nullable = true)
 |    |-- negative: long (nullable = true)
 |    |-- owners: string (nullable = true)
 |    |-- platforms: struct (nullable = true)
 |    |    |-- linux: boolean (nullable = true)
 |    |    |-- mac: boolean (nullable = true)
 |    |    |-- windows: boolean (nullable = true)
 |    |-- positive: long (nullable = true)
 |    |-- price: string (nullable = true)
 |    |-- publisher: string (nullable = true)
 |    |-- release_date: string (nullable = true)
 |    |-

Now we can start unnesting the schema and flattening the dataset.

In [0]:
df = steam.select("*", "data.*").drop("data")

df.printSchema()

root
 |-- appid: long (nullable = true)
 |-- categories: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- ccu: long (nullable = true)
 |-- developer: string (nullable = true)
 |-- discount: string (nullable = true)
 |-- genre: string (nullable = true)
 |-- header_image: string (nullable = true)
 |-- initialprice: string (nullable = true)
 |-- languages: string (nullable = true)
 |-- name: string (nullable = true)
 |-- negative: long (nullable = true)
 |-- owners: string (nullable = true)
 |-- platforms: struct (nullable = true)
 |    |-- linux: boolean (nullable = true)
 |    |-- mac: boolean (nullable = true)
 |    |-- windows: boolean (nullable = true)
 |-- positive: long (nullable = true)
 |-- price: string (nullable = true)
 |-- publisher: string (nullable = true)
 |-- release_date: string (nullable = true)
 |-- required_age: string (nullable = true)
 |-- short_description: string (nullable = true)
 |-- tags: struct (nullable = true)
 |    |-- 1980s: lon

In [0]:
df.limit(5).display()

appid categories ccu developer discount genre header_image initialprice languages name negative owners platforms positive price publisher release_date required_age short_description tags type website 10 List(Multi-player, Valve Anti-Cheat enabled, Online PvP, Shared/Split Screen PvP, PvP) 13990 Valve 0 Action https://cdn.akamai.steamstatic.com/steam/apps/10/header.jpg?t=1666823513 999 English, French, German, Italian, Spanish - Spain, Simplified Chinese, Traditional Chinese, Korean Counter-Strike 5199 10,000,000 .. 20,000,000 List(true, true, true) 201215 999 Valve 2000/11/1 0 Play the world's number 1 online action game. Engage in an incredibly realistic brand of terrorist warfare in this wildly popular team-based game. Ally with teammates to complete strategic missions. Take out enemy sites. Rescue hostages. Your role affects your team's success. Your team's success affects your role. List(266, 1191, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 5426, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 227, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 2784, null, null, null, null, null, null, null, null, null, null, null, null, 1607, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 4831, null, null, null, null, null, null, null, null, null, 1707, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 632, null, null, null, null, null, null, null, null, null, null, null, 3392, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 131, null, null, 769, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 881, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 289, null, null, null, 3353, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 614, null, null, null, null, null, null, 304, null, null, null, 1344, null, null, 1864, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, 1192) game 1000000 List(Single-player, Partial Controller Support, Steam Achievements, Steam Cloud) 0 IndigoBlue Game Studio 0 Action, Adventure, Indie https://cdn.akamai.steamstatic.com/steam/apps/1000000/header.jpg?t=1655723048 999 English, Korean, Simplified Chinese ASCENXION 5 0 .. 20,000 List(false, false, true) 27 999 PsychoFlux Entertainment 2021/05/14 0 ASCENXION is a 2D shoot 'em up game where you explore the field to progress. Players must overcome p

## 1. Macro-level analysis

### 1.1. Which publisher has released the most games on Steam?

In [0]:
df.groupby('publisher').count().sort(F.col('count').desc()).limit(1).show()

+--------------+-----+
|     publisher|count|
+--------------+-----+
|Big Fish Games|  422|
+--------------+-----+



"Big Fish Games" is the publisher who released the most games on Steam, with 442 available games on the platform.

In [0]:
top_publishers = df \
    .filter(df.publisher != "") \
    .groupBy("publisher") \
    .count() \
    .orderBy(F.col("count").desc()) \
    .limit(10)

display(top_publishers)

publisher count Big Fish Games 422 8floor 202 SEGA 165 Strategy First 151 Square Enix 141 Choice of Games 140 Sekai Project 132 HH-Games 132 Ubisoft 127 Laush Studio 126

Databricks visualization. Run in Databricks to view.

### 1.2. What are the best rated games?

On Steam, games ratings are calculated using the approval ratio of users.

In our dataset, each row represents a game, the number of positive and negative reviews are found respectively in the "positive" and "negative" columns.

Therefore, we'll create a "total" column with the total review count, then a "rating" column giving us the positive ratio.

In [0]:
top_rated = df \
    .withColumn("total", F.col("positive") + F.col("negative")) \
    .withColumn("rating", F.round(F.col("positive") / F.col("total"), 3))

In [0]:
top_rated \
    .select(["total", "rating"]) \
    .summary().show()


+-------+------------------+-------------------+
|summary|             total|             rating|
+-------+------------------+-------------------+
|  count|             55691|              55528|
|   mean|1712.7132929916863| 0.7365415285981882|
| stddev| 35687.61118456136|0.24231838102358305|
|    min|                 0|                0.0|
|    25%|                 7|              0.609|
|    50%|                26|              0.799|
|    75%|               145|              0.925|
|    max|           6730438|                1.0|
+-------+------------------+-------------------+



75% of the dataset's games have less than 145 reviews; meanwhile the game which gathered the most reviews reaches a total of 6,730,438!

That is a huge disparity that we should consider when looking at the best rated games : if game A has a 90% approval rating with 10,000 votes and game B has a 100% approval rating but only 100 votes, which game is better? Can a great ratio based on very few reviews be trusted? 

In [0]:
top_rated \
    .select(["name", "total", "rating"]) \
    .orderBy("rating", ascending=False) \
    .limit(20) \
    .display()

name total rating The Spell - A Kinetic Novel 23 1.0 Twisting Mower 1 1.0 Diamonds 4 1.0 Cube Defender 6 1.0 Next Stop 3 17 1.0 Rotund Rebound 5 1.0 N-GON 4 1.0 Zero spring episode 3 1 1.0 Hand of Horzasha 10 1.0 Kooring VR Wonderland:Mecadino's Attack 12 1.0 Magus Over Fool 3 1.0 GUIDE 2 1.0 The Truck Game 1 1.0 Twisty Puzzle Simulator 4 1.0 The Demon Crystal 11 1.0 Symploke: Legend of Gustavo Bueno (Chapter 3) 8 1.0 Fist Of Heaven & Hell 7 1.0 CrossTrix 1 1.0 Kill Tiger 1 1.0 Aether Way 2 1.0

As expected, the games with the best ratings all have 25 reviews or less. We'll therefore set a threshold of 80,000 reviews to rank best rated games.

In [0]:
 top_rated \
    .filter(F.col("total") > 80000) \
    .select(["name", "total", "rating"]) \
    .orderBy("rating", ascending=False) \
    .limit(20) \
    .display()

name total rating People Playground 144569 0.989 Vampire Survivors 131935 0.988 Portal 2 309441 0.988 Hades 202789 0.986 Portal 113538 0.985 RimWorld 144751 0.982 Stardew Valley 506841 0.982 Wallpaper Engine 572127 0.981 Helltaker 108165 0.98 Slime Rancher 99306 0.979 Totally Accurate Battle Simulator 95140 0.979 Slay the Spire 118841 0.979 Terraria 1037091 0.978 The Binding of Isaac: Rebirth 215516 0.977 Mount & Blade: Warband 146087 0.977 Counter-Strike 206414 0.975 Half-Life 2 150137 0.975 Left 4 Dead 2 660664 0.975 Bloons TD 6 216541 0.974 Euro Truck Simulator 2 587983 0.973

Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.

### 1.3. Are there years with more releases? Were there more or fewer game releases during the Covid, for example?

In [0]:
release_date = df.select(F.col("release_date"),F.to_date(F.col("release_date"),"yyyy/M/d").alias("date"))
release_date = release_date.withColumn("year", F.year("date"))

release_year = release_date.groupby("year").count()

display(release_year)

year count 2003 3 2007 98 2018 7663 2015 2566 2006 61 2022 7451 2013 469 null 222 1997 2 2014 1550 2019 6949 2004 6 1998 1 2020 8287 2012 344 2009 309 2016 4176 2001 4 2005 6 2000 2 2010 281 2011 267 2008 159 2017 6006 1999 3 2002 1 2021 8805

Databricks visualization. Run in Databricks to view.

Until 2003, Steam served only as a platform for Valve games, providing updates and multiplayer support. As such, only a handful of games were published of games in its early years.

In 2005, third-party developers were contracted to release games on Steam, thus growing the number of games offered on the platform.

Publishers such as id Software, Eidos Interactive, and Capcom began distributing their games on Steam in 2007, followed by the likes of Ubisoft, THQ, Sega, Take-Two Interactive, Activision, and Electronic Arts in 2008, thus beginning the exponential growth of the service as a digital distribution service.

Since then, only 2019 saw a slowdown in new releases. Many factors could be contributing to this, but one immediately comes to mind : Epic Games Store, a direct competitor, made its debut on December, 6th 2018. Valve, which operates Steam, takes a 30% revenue cut of all games sold through their services, a figure matched by the other services like GOG.com, and console and mobile storefronts. Epic Games opened their digital storefront to challenge Steam by using a 12% revenue split rather than Steam's 30%. This enticed developers and publishers to release their games on Epic Games Store rather than Steam, causing a dip in the amount of games published by Steam the following year.

It doesn't seem Covid (starting in 2020) had a particular effect in the number of releases. One explanation might be that many digital jobs like game developers and publishers are easily performed remotely, and Covid didn't cause too much trouble in the video game market.

### 1.4. How are the prices distributed? Are there many games with a discount?

In [0]:
df.select(["initialprice", "discount", "price"]).summary().show()

+-------+-----------------+------------------+-----------------+
|summary|     initialprice|          discount|            price|
+-------+-----------------+------------------+-----------------+
|  count|            55691|             55691|            55691|
|   mean|797.5663033524268| 2.603777989262179|773.2849832109317|
| stddev|1104.762477841338|12.887080174743176| 1093.13458272345|
|    min|                0|                 0|                0|
|    25%|            199.0|               0.0|            129.0|
|    50%|            499.0|               0.0|            499.0|
|    75%|            999.0|               0.0|            999.0|
|    max|             9999|                90|             9999|
+-------+-----------------+------------------+-----------------+



We easily notice that : 
- "initialprice" and "price" columns must be formatted : the values found are ludicrous for video game prices...unless they are divided by 100 ! For instance, a game where "price" value is 999.0 actually costs 9.99€.
- "discount" is a percentage
- In terms of distribution : 75% games on the platform cost less than 9.99€; less than 25% games have an active discount. 

In [0]:
df = df.withColumns({'initialprice': F.round(F.col("initialprice")/100, 2), 'price': F.round(F.col("price")/100, 2)})

prices = df.select(["initialprice", "discount", "price"])
prices_dist = prices \
    .select(["price"]) \
    .groupby('price') \
    .count() \
    .display()

price count 4.19 11 0.84 3 44.99 44 19.96 1 9.49 1 8.0 3 0.0 7780 4.88 1 1.42 1 12.79 1 7.0 1 249.0 1 84.99 1 4.93 3 3.5 1 8.99 753 10.25 1 16.39 1 6.1 1 0.89 25 1.69 7 8.79 1 124.99 1 3.33 1 3.88 1 8.54 1 3.19 6 9.95 2 7.36 1 2.79 22 1.54 3 31.96 1 1.89 2 24.79 1 19.95 2 2.63 1 8.7 1 5.84 2 89.99 5 15.1 1 2.9 1 27.99 2 6.66 3 5.69 1 39.99 428 2.94 1 9.99 6126 1.94 2 14.99 2878 64.95 1 10.04 3 19.9 2 4.39 4 0.39 7 1.4 1 0.79 40 13.49 9 94.99 4 7.06 1 19.69 1 4.49 41 4.33 1 2.3 1 1.49 73 6.69 7 0.54 21 54.99 15 23.39 1 52.74 1 3.77 2 35.0 1 1.35 1 69.03 1 8.39 5 3.03 1 13.29 2 6.23 1 2.03 1 1.97 3 0.45 2 13.37 1 6.19 1 9.87 1 0.61 2 12.74 5 19.79 1 11.89 1 3.14 1 4.24 4 10.79 4 2.36 1 39.9 1 1.74 12 49.69 1 10.99 430 15.19 2 8.74 1 44.95 1 0.99 5243 8.49 8 2.5 5 149.99 6 5.24 7 1.11 1 1.0 16 39.0 1 1.07 1 13.13 1 24.59 1 7.95 1 269.99 1 6.79 7 11.99 822 2.45 1 2.99 3639 20.99 5 13.99 276 9.74 13 1.88 1 12.59 5 22.99 5 1.99 4193 0.64 3 74.99 3 12.9 1 5.57 1 3.74 29 0.8 1 9.69 1 25.0 3 13.39 1 7.49 40 2.54 2 69.99 25 6.39 6 12.4 1 4.79 18 14.39 1 4.63 1 33.66 1 119.99 2 149.0 1 5.58 1 1.45 1 17.77 1 0.49 303 4.99 6250 7.5 1 4.0 6 1.98 1 5.19 4 26.99 6 3.24 6 5.73 1 17.99 286 7.69 1 2.16 1 37.49 1 2.74 1 3.69 1 11.04 2 4.94 2 18.99 141 9.59 3 6.31 1 13.59 2 8.9 1 18.74 2 6.49 13 19.99 2820 1.87 1 1.95 1 0.57 3 14.5 1 0.5 25 5.39 4 1.58 2 12.95 1 0.75 2 45.0 1 0.62 1 0.38 7 14.49 2 5.1 2 6.99 1856 0.98 1 9.9 6 4.2 8 1.23 2 14.95 1 9.11 1 2.19 14 12.99 796 12.49 14 9.98 1 44.98 1 12.24 1 5.09 10 64.99 6 4.3 1 24.49 1 6.98 1 56.24 1 5.99 1854 2.93 1 9.29 1 11.69 5 29.99 865 25.99 2 20.24 1 3.99 2530 1.79 23 0.51 51 0.76 1 0.67 2 3.09 4 16.66 1 2.39 27 999.0 1 31.99 3 28.99 2 1.19 40 4.89 3 1.24 21 11.49 1 22.49 4 0.56 2 99.99 24 159.99 1 1.39 19 2.14 1 20.04 1 35.99 4 5.65 1 2.59 11 16.24 4 7.99 1580 1.04 6 0.65 1 0.71 1 0.37 19 22.6 1 7.64 1 6.24 5 2.49 99 23.99 9 1.32 1 199.99 17 3.0 9 4.95 1 1.36 1 19.0 2 0.81 2 11.24 7 5.29 1 24.99 862 7.19 5 17.24 1 49.99 190 21.09 1 17.49 7 2.24 12 15.29 3 2.22 1 0.53 5 6.74 4 2.0 8 57.95 1 1.47 2 3.39 6 0.9 32 11.19 1 10.19 3 299.9 1 1.03 4 59.95 1 1.5 6 5.59 9 7.14 2 19.49 2 2.44 4 6.8 1 8.98 1 59.99 187 0.44 1 69.95 1 23.97 1 3.91 1 3.29 2 10.0 10 2.09 34 6.29 9 8.19 1 35.95 1 18.89 1 0.59 97 40.0 1 21.24 1 30.0 1 8.88 1 5.9 1 1.34 3 5.49 6 5.14 1 99.0 1 1.33 2 7.79 7 7.77 1 4.54 3 4.59 1 8.09 1 14.38 1 0.74 34 1.91 1 3.86 1 6.0 5 9.89 7 1.55 1 20.0 5 4.64 1 0.3 2 5.79 1 1.09 5 2.69 8 4.28 1 3.6 1 1.83 1 0.77 4 5.0 20 15.0 9 2.29 1 2.88 1 3.49 45 129.99 5 8.24 4 0.29 11 11.9 1 1.29 8 1.64 1 3.75 1 6.59 2 15.58 1 16.99 185 19.84 1 14.73 1 0.41 10 3.79 2 34.99 155 21.99 9 0.91 4 0.69 38 6.36 1 2.51 1 4.85 1 0.83 2 3.97 2 9.0 2 0.95 1 0.31 6 9.79 3 15.99 230 5.5 1 11.39 1 3.59 20 3.89 8 16.0 1 22.79 1 3.73 1 5.94 2 5.77 1 12.0 4 39.39 1 15.74 1 9.09 3 32.99 1 10.39 4 19.97 1 29.98 2 1.59 25 79.99 13 0.28 20 2.71 4 10.49 9 29.9 1 29.95 1

Since outliers prevent proper visualization, we'll remove them to get a better look at our prices distribution.

In [0]:
prices_dist = prices \
    .where(prices.price<60) \
    .select(["price"]) \

display(prices_dist)

price 9.99 9.99 5.99 19.99 1.99 7.99 12.99 0.0 2.99 13.99 0.99 16.99 0.0 0.0 24.99 4.99 9.99 6.99 0.0 0.0 0.99 0.0 9.99 2.99 19.99 11.99 0.99 14.99 0.99 4.99 7.99 3.99 1.98 6.99 6.99 1.99 19.99 0.99 15.99 14.99 0.0 0.99 12.99 19.99 0.0 10.99 6.99 1.99 6.99 5.99 0.0 4.99 4.99 4.99 9.99 9.99 0.0 2.99 4.99 0.0 16.99 0.0 0.0 9.99 1.99 1.99 0.0 7.99 8.99 12.99 3.99 0.99 14.99 0.0 0.99 14.99 0.99 8.99 2.99 0.0 0.99 3.49 0.0 9.99 0.99 16.99 7.99 19.99 7.99 0.0 0.99 5.99 4.99 2.99 9.99 3.99 3.99 3.99 0.0 1.99 4.99 4.99 0.0 11.99 1.99 3.99 0.0 7.99 0.0 9.99 0.0 0.0 4.99 0.99 2.99 1.99 0.0 0.0 29.99 0.99 14.99 14.99 9.99 5.99 9.99 1.99 9.99 24.99 0.0 14.99 0.99 4.99 1.99 0.99 14.99 39.99 0.99 0.99 6.99 9.99 9.99 3.99 10.99 0.99 7.99 39.99 0.0 4.99 59.99 1.99 2.99 0.99 5.99 16.99 6.99 7.99 0.0 9.99 1.99 4.99 0.99 14.99 2.99 0.0 2.99 0.99 19.99 5.99 4.99 18.99 0.99 2.99 0.99 0.99 4.99 19.99 3.99 0.0 7.99 5.99 12.99 2.99 5.99 14.99 5.99 12.99 12.99 11.99 12.99 0.0 2.99 29.99 3.99 0.0 0.99 14.99 7.99 9.99 15.99 1.5 8.99 24.99 0.0 1.99 2.99 5.99 4.99 0.0 14.99 0.99 0.0 11.99 9.99 0.0 0.0 1.49 24.99 1.99 9.99 0.99 0.99 4.99 0.99 0.0 6.99 4.99 2.99 6.99 6.99 6.99 4.99 0.99 6.99 3.99 0.31 0.0 0.99 24.99 9.99 8.99 2.99 1.99 6.99 0.99 1.99 59.99 1.99 5.99 0.99 3.99 2.99 4.99 0.54 3.99 11.24 0.0 1.99 0.0 9.99 14.99 1.99 0.99 9.99 9.99 3.99 4.99 6.99 0.0 6.99 9.99 0.0 7.99 6.99 5.99 1.99 4.99 4.99 9.99 3.99 0.99 0.0 10.99 1.99 0.99 4.99 9.99 1.99 12.99 0.99 3.99 0.99 1.99 0.99 0.99 9.99 0.99 0.0 9.99 4.89 5.99 0.99 4.99 0.99 1.99 5.59 3.99 2.99 5.99 5.99 0.99 11.99 19.99 4.99 11.99 19.99 0.99 0.0 4.99 5.99 1.99 0.0 1.99 0.9 59.99 0.99 0.0 3.99 9.99 9.99 29.99 29.99 4.99 24.99 6.99 0.99 9.99 0.99 5.99 2.59 0.99 0.0 2.99 0.0 4.99 4.99 0.0 9.99 0.0 0.99 1.99 12.99 7.99 0.41 0.0 3.99 19.99 4.99 14.99 0.99 1.99 1.99 3.99 4.99 3.99 0.0 4.99 0.99 1.99 4.99 9.99 0.99 9.99 7.99 0.99 11.99 1.99 1.99 9.99 1.79 0.0 4.99 4.99 0.0 0.99 1.99 0.0 9.99 1.99 9.99 0.0 0.99 12.99 4.99 0.0 3.99 0.99 10.99 7.99 0.99 14.99 2.99 19.99 0.99 2.99 11.99 39.99 24.99 5.99 2.99 7.99 0.0 6.99 19.99 4.99 9.99 0.0 14.99 4.99 0.99 1.99 9.99 2.99 0.99 5.99 24.99 0.0 5.99 5.99 19.99 9.99 14.99 3.99 0.99 1.99 19.99 0.0 1.99 14.99 5.99 0.0 0.0 4.99 0.99 0.0 0.0 0.44 9.99 34.99 3.99 12.99 4.99 7.99 0.99 4.99 0.64 0.0 29.99 0.99 0.99 24.99 4.99 1.99 9.99 7.99 4.99 19.99 0.99 9.99 4.99 2.99 0.99 0.99 44.99 3.99 3.99 29.99 9.99 13.99 24.99 0.0 0.0 4.49 0.0 5.99 14.99 14.99 7.99 0.99 9.99 1.99 4.99 13.99 2.99 0.0 19.99 19.99 6.99 3.99 1.99 1.99 0.0 14.99 0.99 29.99 5.99 0.0 9.99 0.0 9.99 12.99 3.0 34.99 5.99 9.99 5.99 7.99 3.99 14.99 0.0 3.99 2.99 0.99 0.0 7.99 0.0 2.99 2.99 0.99 7.99 0.0 5.99 2.99 9.99 0.99 2.99 0.99 3.99 19.99 0.99 7.99 9.99 7.99 4.99 0.99 0.99 9.99 0.0 9.99 7.99 0.0 2.99 19.99 1.99 3.99 11.99 6.99 5.99 6.99 9.99 4.99 4.99 0.99 0.99 4.99 0.99 7.99 2.49 9.99 0.0 12.99 14.99 12.99 6.49 19.99 0.0 0.99 6.99 19.99 9.99 19.99 9.99 14.99 14.99 14.99 0.0 3.99 19.99 4.99 19.99 14.99 11.99 4.99 9.9 4.99 0.0 59.99 12.99 0.28 0.0 4.99 0.0 0.0 4.99 2.99 14.99 1.99 4.99 0.99 15.99 9.99 9.99 0.0 11.99 19.99 3.99 4.99 12.99 19.99 29.99 0.0 3.99 0.0 5.09 14.99 2.99 19.99 9.99 15.99 29.99 0.0 19.99 2.99 0.0 0.0 1.99 24.99 11.99 2.99 2.99 24.99 1.99 9.99 9.99 6.99 0.0 11.99 0.99 1.99 2.99 4.99 1.99 6.99 2.99 3.99 2.99 4.99 2.99 5.99 5.99 19.99 1.49 19.99 19.99 19.99 19.99 1.99 11.99 0.99 13.99 4.99 29.99 3.99 5.99 6.99 18.99 0.99 9.99 0.0 7.99 4.99 19.99 0.0 0.0 14.99 14.99 4.99 3.99 14.99 11.69 14.99 14.99 3.99 0.0 17.99 9.99 4.99 11.99 19.99 0.99 3.99 9.99 4.2 0.99 5.99 14.99 0.0 4.99 4.99 14.99 0.0 8.99 9.99 17.99 17.99 19.99 1.99 11.99 0.99 4.99 4.99 8.99 0.0 0.0 29.99 1.99 19.99 4.99 0.99 4.99 19.99 12.99 11.99 8.99 9.99 1.99 14.99 9.99 0.99 29.99 9.99 19.99 0.0 0.0 0.99 11.99 6.99 19.99 0.99 1.99 0.0 39.99 16.99 1.99 9.99 0.99 0.0 1.99 4.99 11.99 0.0 13.99 5.99 0.99 6.99 1.99 9.99 0.99 4.94 19.99 0.99 15.99 1.99 1.99 0.99 4.99 2.99 0.99 10.99 9.99 1.99 4.99 0.59 12.99 0.99 2.99 2.99 1

Databricks visualization. Run in Databricks to view.

In [0]:
discount = df \
    .withColumn("is_discounted", F.when(df.discount == 0, "No discount").otherwise("Discounted")) \
    .groupby("is_discounted") \
    .count()
 
display(discount)

is_discounted count Discounted 2518 No discount 53173

Databricks visualization. Run in Databricks to view.

### 1.5. What are the most represented languages?

In [0]:
languages = df \
    .withColumn('language', F.explode(F.split(F.col("languages"), ", "))) \
    .groupBy('language') \
    .count() \
    .orderBy(F.col('count').desc())

languages.limit(3).display()

language count English 55116 German 14019 French 13426

Databricks visualization. Run in Databricks to view.

Unsurprisingly, English is by far the most represented language. Following are German in second and French in third, with close counts. 

### 1.6. Are there many games prohibited for children under 16/18?

In [0]:
age_restricted_16 = df \
    .filter(df["required_age"] >= 16) \
    .count()
age_restricted_18 = df \
    .filter(df["required_age"] >= 18) \
    .count()
print(f"There are {age_restricted_16} games in the dataset that are prohibited for children under 16, and {age_restricted_18} games forbidden to people under 18.")

There are 305 games in the dataset that are prohibited for children under 16, and 229 games forbidden to people under 18.


**PEGI-16** : 305 games

**PEGI-18** : 229 games

## 2. Genres analysis

###  2.1. What are the most represented genres?

In [0]:
df = df.withColumn("genre", F.explode(F.split(F.col("genre"), ", ")))

genres = df \
    .filter(df.genre != "") \
    .groupBy("genre") \
    .count() \
    .orderBy(F.col("count") \
    .desc()) \
    .limit(15)

display(genres)

genre count Indie 39681 Action 23759 Casual 22086 Adventure 21431 Strategy 10895 Simulation 10836 RPG 9534 Early Access 6145 Free to Play 3393 Sports 2666 Racing 2155 Massively Multiplayer 1460 Utilities 682 Design & Illustration 406 Animation & Modeling 322

Databricks visualization. Run in Databricks to view.

###  2.2. Are there any genres that have a better positive/negative review ratio?

In [0]:
genres_ratio = df \
    .filter(df.genre != "") \
    .select(["genre", "positive", "negative"]) \
    .groupby("genre") \
    .sum("positive", "negative") \
    .withColumn("ratio", F.round(F.col("sum(positive)") / F.col("sum(negative)"), 2)) \
    .select(F.col("genre"), F.col("ratio")) \
    .orderBy(F.col("ratio") \
    .desc()) \
    .limit(20)

display(genres_ratio)

genre ratio Photo Editing 42.03 Animation & Modeling 26.17 Design & Illustration 24.96 Utilities 17.0 Game Development 8.39 Indie 7.67 Audio Production 7.33 Video Production 6.82 Casual 6.53 Web Publishing 6.49 Simulation 6.49 Racing 6.1 RPG 5.93 Action 5.66 Strategy 5.6 Adventure 5.25 Early Access 4.63 Gore 4.44 Free to Play 4.37 Sports 4.08

Databricks visualization. Run in Databricks to view.

###  2.3. Do some publishers have favorite genres?

In [0]:
pub_count = df \
    .filter(df.publisher != "") \
    .groupBy('publisher') \
    .agg(F.count(F.col('publisher')).alias('total_count'))
    
pub_genres = df \
    .filter(df.genre != "") \
    .join(pub_count, df['publisher'] == pub_count['publisher'],'inner') \
    .select(df['*'], pub_count['total_count']) \
    .groupBy(['publisher','genre','total_count']) \
    .agg(F.count(F.col('genre')).alias('genre_count')) \
    .withColumn('proportion', F.round(F.col('genre_count')/F.col('total_count'), 2)) \
    .orderBy(['proportion'], ascending = False)

display(pub_genres.filter(pub_genres.total_count > 10))

publisher genre total_count genre_count proportion id Software Action 16 16 1.0 PopCap Games, Inc. Casual 19 19 1.0 HexWar Games Strategy 43 42 0.98 CITY CONNECTION Action 19 18 0.95 grin robot Indie 11 10 0.91 HeR Interactive Adventure 33 29 0.88 Notus Games Ltd Casual 13 11 0.85 8floor Casual 243 202 0.83 Sandlot Games Casual 11 9 0.82 SNK CORPORATION Action 30 24 0.8 Dogenzaka Lab Adventure 19 15 0.79 Ziggurat, 612 Games Action 14 11 0.79 Arc System Works Action 36 27 0.75 Remix Games Casual 12 9 0.75 Sigyaad Team Casual 12 9 0.75 WASABI entertainment RPG 23 17 0.74 Mens Sana Interactive Casual 27 20 0.74 PopCap Games, Inc., Electronic Arts Casual 15 11 0.73 CI Games Action 14 10 0.71 Telltale Games Adventure 21 15 0.71 Interplay Inc. Action 16 11 0.69 rondomedia GmbH Simulation 13 9 0.69 SEGA, Feral Interactive (Mac), Feral Interactive (Linux) Strategy 19 13 0.68 MangaGamer Adventure 66 45 0.68 Zojoi Adventure 12 8 0.67 Skunkape Games Adventure 24 16 0.67 CAPCOM Co., Ltd. Action 31 20 0.65 Reforged Group Indie 135 88 0.65 HotFoodGames Indie 26 17 0.65 My Label Game Studio Indie 34 22 0.65 Shaman Games Studio Casual 22 14 0.64 BJGameInd Indie 11 7 0.64 DigiMight Casual 25 16 0.64 Robotizar Games Casual 11 7 0.64 Notovia Indie 11 7 0.64 HUNTERS Action 14 9 0.64 InvertMouse Indie 16 10 0.63 Skybound Games Adventure 16 10 0.63 NovaLogic, THQ Nordic Action 16 10 0.63 SosiskaGames Indie 24 15 0.63 Argent Games Casual 13 8 0.62 Boomzap Inc Casual 21 13 0.62 Rockstar Games Action 26 16 0.62 Slitherine Ltd. Strategy 157 96 0.61 Apogee Entertainment Action 23 14 0.61 Valve Action 45 27 0.6 Sprovieri Games Casual 20 12 0.6 Rainbow Games Casual 17 10 0.59 LoadUpGames.com Casual 17 10 0.59 INTI CREATES CO., LTD. Action 22 13 0.59 H2 Interactive Co., Ltd. Action 17 10 0.59 HIKARI FIELD Adventure 32 19 0.59 William at Oxford Casual 88 51 0.58 SimBin Racing 12 7 0.58 JustE Publishing Indie 12 7 0.58 Cosmi Valusoft Simulation 23 13 0.57 Humongous Entertainment, Nightdive Studios Casual 56 32 0.57 The Revills Games Casual 14 8 0.57 iWin Casual 21 12 0.57 3D Realms (Apogee Software) Action 28 16 0.57 City Interactive S.A. Adventure 16 9 0.56 Cronostase Casual 16 9 0.56 familyplay Casual 63 35 0.56 Shiravune Adventure 29 16 0.55 Techland Publishing Action 11 6 0.55 Mastertronic Indie 11 6 0.55 Zoo Corporation Casual 55 30 0.55 Pokies4fun Simulation 11 6 0.55 Introversion Software Indie 11 6 0.55 Clouded Leopard Entertainment RPG 20 11 0.55 Skull Box Games Indie 11 6 0.55 MumboJumbo Casual 62 34 0.55 Revolution Software Ltd Adventure 13 7 0.54 Kavkaz Sila Games Simulation 26 14 0.54 Spike Chunsoft Co., Ltd. Adventure 41 22 0.54 KoheiGallery Indie 13 7 0.54 Frogwares Adventure 26 14 0.54 ImperiumGame Indie 75 40 0.53 Google Inc. Adventure 17 9 0.53 Renan Miguelote Vianna Casual 15 8 0.53 PlayFirst Casual 17 9 0.53 Microïds Indie Adventure 21 11 0.52 Mediascape Co., Ltd. Action 33 17 0.52 Capcom Action 62 32 0.52 Vidas Salavejus Indie 21 11 0.52 Bethesda Softworks Action 71 37 0.52 LucasArts, Lucasfilm, Disney Action 35 18 0.51 8Floor Casual 80 41 0.51 IR Studio Indie 12 6 0.5   Gales Corp.    Indie 16 8 0.5 Flying Islands Team Indie 12 6 0.5 AMG GAMES Action 12 6 0.5 Hunny Bunny Studio Indie 28 14 0.5 Hunny Bunny Studio Casual 28 14 0.5 Retroism Simulation 16 8 0.5 Do Games Limited Casual 12 6 0.5 Do Games Limited Adventure 12 6 0.5 Cute Hannah's Games Casual 48 24 0.5 SCS Software Simulation 12 6 0.5 SCS Software Indie 12 6 0.5 Big Fish Games Casual 833 418 0.5 Grey Alien Games Casual 48 24 0.5 DIG Publishing Indie 20 10 0.5 DIG Publishing Casual 20 10 0.5 Khud0 Indie 14 7 0.5 Pirotexnik Indie 16 8 0.5 Pirotexnik Casual 16 8 0.5 Cyan Worlds Inc Adventure 20 10 0.5 Polygon Art Indie 12 6 0.5 Lingtan Studio Adventure 12 6 0.5 Sweety Boy Casual 14 7 0.5 Sweety Boy Indie 14 7 0.5 FireFly Studios Strategy 20 10 0.5 Wadjet Eye Games Adventure 30 15 0.5 Paul Connor Action 26 13 0.5 Avalon Digital Strategy 30 15 0.5 DominiGames Casual 26 13 0.5 Domini

Databricks visualization. Run in Databricks to view.

###  2.4. What are the most lucrative genres?

In [0]:
sales = df \
    .filter(df.genre != "") \
    .groupBy('genre') \
    .agg(F.round(F.sum("price")).alias("total_sales")) \
    .orderBy(F.col('total_sales').desc()) \
    .limit(10)

display(sales)

genre total_sales Indie 260630.0 Action 183588.0 Adventure 171582.0 Casual 123836.0 Simulation 98517.0 Strategy 91572.0 RPG 86213.0 Early Access 53758.0 Sports 23855.0 Racing 17716.0

Databricks visualization. Run in Databricks to view.

In [0]:
price_per_game = df \
    .filter(df.genre != "") \
    .groupBy("genre") \
    .agg(F.round(F.mean("price"), 2) \
    .alias("avg_price_per_game")) \
    .orderBy(F.desc("avg_price_per_game")) \

display(price_per_game)

genre avg_price_per_game Web Publishing 21.8 Game Development 21.28 Photo Editing 20.33 Audio Production 19.64 Design & Illustration 19.06 Software Training 19.03 Video Production 18.96 Animation & Modeling 18.78 Accounting 14.43 Education 14.34 Utilities 11.45 Simulation 9.09 RPG 9.04 Sports 8.95 Early Access 8.75 Strategy 8.4 Racing 8.22 Adventure 8.01 Action 7.73 Nudity 6.96 Sexual Content 6.88 Indie 6.57 Gore 5.7 Casual 5.61 Violent 5.28 Massively Multiplayer 5.0 Free to Play 0.29 Movie 0.0

Databricks visualization. Run in Databricks to view.

It's worth noting that while indie may be the most profitable genre, indie games individual prices tend to be on the lower side . The indie genre represents many sales, simply because it is by far the most represented genre on the platform.

## 3. Platform analysis

###  3.1. Are most games available on Windows/Mac/Linux instead?

In [0]:
df = df \
    .withColumn("Linux", F.col("platforms")["linux"].cast('int')) \
    .withColumn( "MacOS", F.col("platforms")["mac"].cast('int')) \
    .withColumn("Windows", F.col("platforms")["windows"].cast('int'))

counts = df \
    .agg(F.sum(F.col("Linux")).alias("Linux"),
         F.sum(F.col('MacOS')).alias("MacOS"),
         F.sum(F.col('Windows')).alias("Windows"))
    
display(counts)

Linux MacOS Windows 23989 35902 157076

Databricks visualization. Run in Databricks to view.

In [0]:
pd_counts = counts.toPandas().transpose().reset_index().rename(columns={"index":'platform', 0:"available_games"})
t_counts = spark.createDataFrame(pd_counts)
tot = df.count()
platform_perc = t_counts \
    .withColumn("proportion_available", F.round(F.col("available_games")/tot * 100, 2))
display(platform_perc)

platform available_games proportion_available Linux 23989 15.27 MacOS 35902 22.85 Windows 157076 99.98

Databricks visualization. Run in Databricks to view.

###  3.2. Do certain genres tend to be preferentially available on certain platforms?

In [0]:
genre_platforms = df \
    .groupBy("genre") \
    .agg(F.sum(F.col("Linux")).alias("Linux"),
         F.sum(F.col("MacOS")).alias("Mac"),
         F.sum(F.col("Windows")).alias("Windows"))

display(genre_platforms)

genre Linux Mac Windows Education 19 56 317 Massively Multiplayer 164 270 1459 Sexual Content 7 13 54 Adventure 3302 5039 21427 Sports 287 506 2665 Accounting 0 4 16 Audio Production 7 41 193 Video Production 6 29 247 Animation & Modeling 38 74 322 Racing 304 424 2154 Gore 14 20 99 Design & Illustration 54 100 405 Software Training 12 27 164 Web Publishing 8 20 89 Photo Editing 4 16 105 Utilities 49 102 681 Violent 22 33 168 Game Development 35 52 159 Movie 0 0 1 Nudity 7 10 45 Early Access 632 900 6145 Casual 3305 5130 22082 Action 3379 4564 23755 Strategy 1826 3005 10892 Indie 6978 9935 39676 Free to Play 474 845 3391 RPG 1524 2248 9533 Simulation 1532 2439 10832

Databricks visualization. Run in Databricks to view.

## Conclusion

Performing this analysis has given us a detailed view of Steam's ecosystem and a better understanding of the platform. 

It is crucial for publishers to finely tune their products to a demanding consumer base's expectations. Given Steam's domineering position in the computer games market, these findings can help publishers make strategic decisions such as : 
- when to release a game 
- what genres are currently in demand
- how to price their products
- who their direct competitors are
- whether porting to several platforms would be beneficial

and more!